src/core/DataFile.cpp

```diff
@@ -89,18 +89,17 @@ const std::vector<ProjectVersion> DataFile::UPGRADE_VERSIONS = {
 	"1.2.0-rc3"        ,   "1.3.0"
 };
 
-DataFile::typeDescStruct
-		DataFile::s_types[DataFile::TypeCount] =
+std::array<DataFile::typeDescStruct, DataFile::TypeCount> DataFile::s_types =
```

Comment:
Can you elaborate on this? Omitting the template arguments here gives a deduction error. The same occurs for ``InstrumentFunctions`` as well.
``no viable constructor or deduction guide for deduction of template arguments of array``

I'm not sure how I would go about deducing these types in this instance (or if it is necessary).

Reply:
It should work if you construct each array element with `DataFile::typeDescStruct`, rather than having raw initialiser lists (IIRC this was how you had it when I wrote that comment). I don't mind whether you put the type in the template parameters or the elements of the initialiser list; usually I prefer to have it on the elements, as then you can also infer the array size, but it does make it somewhat more verbose in this case.

You also need to remove the declarations from the header files when using `auto` to avoid getting redefinition errors. The variables in question are only used in their own source files, so there's no need to forward-declare them in headers (which forces you to specify the type explicitly, as there's no definition to deduce it from). They can be static variables in the source file only (even better, [define them in an anonymous namespace](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sf22-use-an-unnamed-anonymous-namespace-for-all-internalnon-exported-entities)).